In [2]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

df_fot = pd.read_csv(r"C:\Users\carla\Downloads\Iron Hack  - Curso\projcet-3-IronHack Delivery\fake_orders_test.csv")

df_fot

order_id       activation_time_local  country_code  store_address  \
0      33557880  2019-03-10 23:59:59.000000            AR          14200   
1      33512615  2019-03-10 23:58:32.000000            TR          28725   
2      33512451  2019-03-10 23:57:56.000000            TR          28725   
3      33530892  2019-03-10 23:57:33.000000            ES          19777   
4      33557765  2019-03-10 23:57:21.000000            AR          34565   
...         ...                         ...           ...            ...   
60395  31960607  2019-03-01 00:04:31.000000            TR          68820   
60396  32002079  2019-03-01 00:03:53.000000            AR          50175   
60397  32002046  2019-03-01 00:03:06.000000            AR          55159   
60398  32001950  2019-03-01 00:01:01.000000            AR          62504   
60399  32001318  2019-03-01 00:01:00.000000            AR          55159   

          final_status payment_status  products  products_total  \
0      DeliveredStatus           PAID         1            4.54   
1      DeliveredStatus           PAID         1            3.76   
2      DeliveredStatus           PAID         1            2.86   
3       CanceledStatus           PAID         1           12.95   
4      DeliveredStatus           PAID         2            2.86   
...                ...            ...       ...             ...   
60395  DeliveredStatus           PAID         2           10.17   
60396   CanceledStatus           PAID         1            3.80   
60397  DeliveredStatus           PAID         1            4.24   
60398   CanceledStatus           PAID         1            4.91   
60399  DeliveredStatus       NOT_PAID         2            5.14   

       purchase_total_price  
0                      8.64  
1                      3.76  
2                      2.86  
3                      0.00  
4                      6.48  
...                     ...  
60395                 10.33  
60396                  0.00  
60397                  4.91  
60398                  0.00  
60399                  5.58  

[60400 rows x 9 columns]

In [3]:
df_fot.head()

order_id       activation_time_local  country_code  store_address  \
0  33557880  2019-03-10 23:59:59.000000            AR          14200   
1  33512615  2019-03-10 23:58:32.000000            TR          28725   
2  33512451  2019-03-10 23:57:56.000000            TR          28725   
3  33530892  2019-03-10 23:57:33.000000            ES          19777   
4  33557765  2019-03-10 23:57:21.000000            AR          34565   

      final_status payment_status  products  products_total  \
0  DeliveredStatus           PAID         1            4.54   
1  DeliveredStatus           PAID         1            3.76   
2  DeliveredStatus           PAID         1            2.86   
3   CanceledStatus           PAID         1           12.95   
4  DeliveredStatus           PAID         2            2.86   

   purchase_total_price  
0                  8.64  
1                  3.76  
2                  2.86  
3                  0.00  
4                  6.48

**1. KPI**

**KPI 1.Promedio de productos por envio segmentado por store_address**

Este KPI es útil para entender la capacidad de cada tienda en términos de volumen de productos gestionados por pedido.
Para calcular el promedio de productos por envío segmentado por store, utilizamos el campo store_address (que representa la tienda) y el campo products (que indica el número de productos en cada pedido).

**KPI 2. Importe promedio por envio segmentado por country_coe**e
Esta métrica muestra el promedio del importe por pedido en cada país, permitiéndonos entender el valor promedio de los pedidos según la localización.

**KPI 3.: Ratio de Canceled Orders per Delivered Orde**

Este ratio ayuda a evaluar la proporción de pedidos cancelados en comparación con los entregados.
rs

**4. Análisis Exploratorio de Datos**

**4.1. Que porcentaje de pedidos estan autorizados?**

In [8]:
#Crear una columna para identificar pedidos subautorizados
df_fot['subauthorized'] = df_fot['products_total']< df_fot['purchase_total_price']
df_fot

order_id       activation_time_local  country_code  store_address  \
0      33557880  2019-03-10 23:59:59.000000            AR          14200   
1      33512615  2019-03-10 23:58:32.000000            TR          28725   
2      33512451  2019-03-10 23:57:56.000000            TR          28725   
3      33530892  2019-03-10 23:57:33.000000            ES          19777   
4      33557765  2019-03-10 23:57:21.000000            AR          34565   
...         ...                         ...           ...            ...   
60395  31960607  2019-03-01 00:04:31.000000            TR          68820   
60396  32002079  2019-03-01 00:03:53.000000            AR          50175   
60397  32002046  2019-03-01 00:03:06.000000            AR          55159   
60398  32001950  2019-03-01 00:01:01.000000            AR          62504   
60399  32001318  2019-03-01 00:01:00.000000            AR          55159   

          final_status payment_status  products  products_total  \
0      DeliveredStatus           PAID         1            4.54   
1      DeliveredStatus           PAID         1            3.76   
2      DeliveredStatus           PAID         1            2.86   
3       CanceledStatus           PAID         1           12.95   
4      DeliveredStatus           PAID         2            2.86   
...                ...            ...       ...             ...   
60395  DeliveredStatus           PAID         2           10.17   
60396   CanceledStatus           PAID         1            3.80   
60397  DeliveredStatus           PAID         1            4.24   
60398   CanceledStatus           PAID         1            4.91   
60399  DeliveredStatus       NOT_PAID         2            5.14   

       purchase_total_price  subauthorized  
0                      8.64           True  
1                      3.76          False  
2                      2.86          False  
3                      0.00          False  
4                      6.48           True  
...                     ...            ...  
60395                 10.33           True  
60396                  0.00          False  
60397                  4.91           True  
60398                  0.00          False  
60399                  5.58           True  

[60400 rows x 10 columns]

In [9]:
#Calcular el porcentaje de pedidos subautorizados
subauthorized_percentage = (df_fot['subauthorized'].sum()/len(df_fot))*100
print(f"Porcentaje de pedidos subautorizados: {subauthorized_percentage:.2f}%")

Porcentaje de pedidos subautorizados: 57.58%


**4.2. Qué porcentajes de pedidos se autorizan correctamente con una autorización incremental (+20%) sobre el monto en el checkout?**

In [11]:
#Crear una columna para la autorización del 20%
df_fot['20_increase'] = df_fot['products_total'] * 1.20
#Verificar si la autorizacion incremental cubre el precio de compra
df_fot['is_authorized'] = df_fot['20_increase'] >= df_fot['purchase_total_price']
#Calcular el porcentaje de pedidos correctamente autorizados con una autorización incremental del 20%.
correctly_authorized_percentage = (df_fot['is_authorized'].sum()/len(df_fot)) * 100
print(f'Porcentaje de pedidos correctamente autorizados con autorización incremental (20 %): {correctly_authorized_percentage:.2f}%')

Porcentaje de pedidos correctamente autorizados con autorización incremental (20 %): 69.03%


**4.3. Hay diferencias cuando se dividen por país?**

In [13]:
#### Calcular el % de pedidos subautorizados y correctamente autorizados por país.
country_stats = df_fot.groupby(' country_code').agg(
    total_orders=('order_id','count'),
    subauthorized_orders=('subauthorized','sum'),
    correctly_authorized_orders=('is_authorized','sum'))
#Calcular los %
country_stats['subauthorized_percentage'] = (country_stats['subauthorized_orders']/country_stats['total_orders'])*100
country_stats['correctly_authorized_percentage'] = (country_stats['correctly_authorized_orders']/country_stats['total_orders'])*100

print(country_stats[['subauthorized_percentage','correctly_authorized_percentage']])

               subauthorized_percentage  correctly_authorized_percentage
 country_code                                                           
AR                            60.597332                        66.388720
BR                            50.000000                        50.000000
CI                            33.333333                        66.666667
CL                            87.047101                        12.952899
CR                            44.896116                        79.855465
DO                            63.600000                        62.000000
EC                            64.234450                        62.599681
EG                            62.910284                        65.645514
ES                            62.197407                        71.497735
FR                            65.636704                        67.322097
GE                            51.831502                        68.315018
GT                            33.156966            

Conclusión: Si, hay diferencias significativas cuando se dividen los pedidos por país. Los % de pedidos subautorizados y correctamente autorizados varían notablemente según la región, lo cual sugiere una fluctuación en la relación entre el product_total y el purchase_total_price.

**4.4. Para el resto de pedidos que quedaria fuera de la autorización incremental, ¿qué valores serían necessarios para capturar el monto restante?**

In [17]:
#Filtrar los pedidos donde la autorización incremental no es suficiente.
not_authorized = df_fot[df_fot['is_authorized'] == False]
#Calcular el valor necesario para capturar el monto restante
df_fot['remaining_amount'] = (not_authorized['products_total']*1.2) - not_authorized['purchase_total_price']
#Ver el monto restante
print(not_authorized[['order_id','remaining_amount']].head())

    order_id  remaining_amount
0   33557880            -3.192
4   33557765            -3.048
10  33523444            -7.710
12  33530379            -0.480
20  33509944            -0.426


**4.5. ¿Qué tiendas son las más problemáticas en términos de pedidos y valor monetario?**

In [19]:
#Agrupar por tienda y calcular el numero de pedidos y el valor total de los pedidos subautorizados porque son los que tienen más discrepancias y por tanto seran más problemaáticas
tiendas = df_fot.groupby('store_address').agg(
    num_orders=('order_id','count'),
    subauthorized_values = ('purchase_total_price', lambda x: x[df_fot['subauthorized']].sum())
).sort_values(by='subauthorized_values', ascending=False)
top_5_tiendas = tiendas.head(5)
print(top_5_tiendas)

               num_orders  subauthorized_values
store_address                                  
28669                 201               5832.25
28671                 492               4536.10
28712                 242               3575.47
12513                 276               2933.10
14455                 252               2924.07


**4.6. Para pedidos subautorizados, ¿hay una correlación entre la diferencia en los precios y la cancelación del pedido? En otras, palabras:
¿Es más probable que se cancele el pedido a medida que aumenta la diferencia de precio?**

In [20]:
#Crear una columna con la diferencia de precios
df_fot['difference_prices'] = df_fot['purchase_total_price'] - df_fot['products_total']
df_fot['canceled'] = (df_fot['final_status'] == 'CanceledStatus').astype(int)

#Correlación entre al diferencia de precios y la cancelación
correlation = df_fot[['difference_prices','canceled']].corr()
print(correlation)

                   difference_prices  canceled
difference_prices           1.000000 -0.432955
canceled                   -0.432955  1.000000


Conclusión: La correlación indica que cuanto mayor es la discrepancia en los precios, más probable es que el pedido sea cancelado. 

**3. Experimento**

**3.1. Tipo de test**

Realizariamos una test A/B donde dividiria a los usuarios en dos grupos aleatorios:
- Grupo de Control: Tarifa de entrega actual de 1.9€
- Grupo de Experimento: Nueva tarifa de entrega de 2.1€

**3.2. Público del test**

Aplicaria el test a todos los usuarios activos, ya que el impacto de un tarifa de entrega afecta la decisión de compra de ambos segmentos y 
podría tener diferentes efectos en cada grupo.

**3.3. Suposiciones**

- Precio: Asumimos que el incremento de la tarifa podria reducir la frecuencia de pedidos o el gasto promedio por cliente.
- Comportamiento Nuevos vs Recurrentes: Los usuarios recurrentes pueden ser menos sensibles al cambio, mientras que los nuevos podrian ser más 
propensos a cancelar el pedido.

**3.3. Validación**

Usar datos historicos para realizar analisis deL precio antes de implementar el experimento, para tener una idea inicial de cómo
podria responder cada grupo.

**3.4. Determinación de la duración del experimento**

Para determinar la muestra del experimento realizaremos divisiones aleatorias para prever el EMD en cada nivel. 

**3.5. KPIs para evaluar el exito**

- Pedidos
- Ingreso promedio por pedido
- Valor promedio de producto por pedido
- Margen de tarifa relativa

**3.6. Análisis de resultados**

Realizar pruebas para comparar cada KPI en ambos grupos. Usar t-tests, para ver si existe 
    una diferencia significativa entre los dos grupos.


**3.7. Recomendaciones**
- Si el aumento de la tarifa no afecta significativamente los KPIs clave o tiene un impacto positivo, 
recomendaria implementar el aumento de tarifa de 2.1€
- Si el aumento reduce significativamente la frecuencia de pedidos, recomendaria:
    - Segmentar la tarifa: Aplicar el aumento solo para ciertos segmentos menos sensibles al precio, como usuarios con alta frecuencia de pedidos.
    - Ofrecer incentivos a nuevos usuarios: Si los usuarios nuevos muestran una mayor sensibilidad al aumento, considerar tarifas promocionales
    podria ayudar a mejorar la tasa de retención.

In [47]:
#Division los usuarios en dos grupos aleatorios y asignar al grupo de control la tarifa 1.9€ y al grupo experimento 2.1€
df_fot['grupo'] = np.random.choice(['control','experimento'], size=len(df_fot), p = [0.5, 0.5])
df_fot['delivery_fee'] = np.where(df_fot['grupo'] == 'control', 1.9, 2.1)

In [70]:
#Definir KPIs

#Ingreso promedio por pedido
average_revenue_control = df_fot[df_fot['grupo'] == 'control']['purchase_total_price'].mean()
average_revenue_experimento = df_fot[df_fot['grupo'] == 'experimento']['purchase_total_price'].mean()

#Margen de tarifa relativa
df_fot['relative_margin'] = ((df_fot['purchase_total_price'] - df_fot['products_total']) / df_fot['products_total']) * 100
relative_margin_control = df_fot[df_fot['grupo'] == 'control']['relative_margin'].mean()
relative_margin_experimento = df_fot[df_fot['grupo'] == 'experimento']['relative_margin'].mean()

#Valor promedio de producto por pedido
df_fot['average_product_value'] = df_fot['purchase_total_price']/df_fot['products']
average_product_value_control = df_fot[df_fot['grupo'] == 'control']['average_product_value'].mean()
average_product_value_experimento = df_fot[df_fot['grupo'] == 'experimento']['average_product_value'].mean()

In [72]:
#Pruebas estadisiticas entre grupos
from scipy.stats import ttest_ind, ks_2samp
#T-test para comparar ingresos promedio
t_stat_revenue, p_value_revenue = ttest_ind(
    df_fot[df_fot['grupo'] == 'control']['purchase_total_price'],
    df_fot[df_fot['grupo'] == 'experimento']['purchase_total_price'],
    equal_var=False)

#T-test para comparar el Margen de tarifa relativa
t_stat_margin, p_value_margin = ttest_ind(
    df_fot[df_fot['grupo'] == 'control']['relative_margin'],
    df_fot[df_fot['grupo'] == 'experimento']['relative_margin'],
    equal_var= False)

#T-test para comparar el Valor promedio de producto por pedido
t_stat_product_value, p_value_product_value = ttest_ind(
    df_fot[df_fot['grupo'] == 'control']['average_product_value'],
    df_fot[df_fot['grupo'] == 'experimento']['average_product_value'],
    equal_var=False)

In [74]:
#Resultados

results = {
    'Ingreso Promedio por Pedido': {'control_mean': average_revenue_control, 'experimento_mean': average_revenue_experimento, 'p_value': p_value_revenue},
    'Margen de Tarifa Relativa': {'control_mean': relative_margin_control, 'experimento_mean': relative_margin_experimento, 'p_value': p_value_margin},
    'Valor Promedio de Productos por Pedido': {'control_mean': average_product_value_control, 'experimento_mean': average_product_value_experimento, 'p_value': p_value_product_value},
}

# Filtrar las métricas que resultaron significativas (p-value < 0.05)
significant_results = {metric: stats for metric, stats in results.items() if stats['p_value'] < 0.05}
results, significant_results

({'Ingreso Promedio por Pedido': {'control_mean': 10.69516768292683,
   'experimento_mean': 10.711713869772368,
   'p_value': 0.8504161722188557},
  'Margen de Tarifa Relativa': {'control_mean': inf,
   'experimento_mean': 258.89860094884017,
   'p_value': nan},
  'Valor Promedio de Productos por Pedido': {'control_mean': 5.185869966116287,
   'experimento_mean': 5.112810435629457,
   'p_value': 0.15989468220267983}},
 {})

**Conclusiones:**

**Ingreso Promedio por Pedido**

* Media grupo control: 10.70
* Media grupo experimento: 10.71
* p-valor: 0.8504 (no significativo)

La diferencia en el ingreso promedio por pedido entre los grupos no es significativa. Esto indica que el aumento de la tarifa no tuvo un impacto
notable en el gasto promedio de los pedidos.

**Valor Promedio de Productos por Pedido**

* Media grupo control: 5.19
* Media grupo experimento: 5.11
* p-valor: 0.1599 (no significativo)

La diferencia en el valor de productos por pedido tampoco es significativa. Esto sugiere que el aumento de la tarifa de entrega no tuvo un impacto
notable.